snuba/utils/streams/abstract.py

```diff
@@ -0,0 +1,42 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Callable, Generic, Mapping, Optional, Sequence, TypeVar
+
+
+TStream = TypeVar("TStream")
+
+TOffset = TypeVar("TOffset")
+
+TValue = TypeVar("TValue")
+
+
+@dataclass
+class Message(Generic[TStream, TOffset, TValue]):
+    stream: TStream
+    offset: TOffset
+    value: TValue
+
+
+class Consumer(ABC, Generic[TStream, TOffset, TValue]):
+    @abstractmethod
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TStream]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[TStream]], None]] = None,
+    ) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def poll(
+        self, timeout: Optional[float] = None
+    ) -> Optional[Message[TStream, TOffset, TValue]]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def commit(self) -> Mapping[TStream, TOffset]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def close(self) -> None:
+        raise NotImplementedError
```

Comment:
It's been a while, but this API should hopefully look a _little_ familiar: https://github.com/getsentry/snuba/blob/aaef1c6c3e648531132743f695a6d14e80baf9e4/snuba/subscriptions/consumer.py#L132-L406